File: src/completed/python/euler0011.py
###
#
# Created by Holden on 12/17/2015
#
# SOLVED on 12/17/2015
#
# Problem 11 - Largest Product in a Grid
#
###

def greatest_product(mult, filename):
    import math
    new_list = []
    product = 1
    
    x20 = open(filename, "r")

    for alist in x20:
        element = alist.split()
        for i in range(20):
            new_list.append(int(element[i].strip()))
            
    index = int(math.sqrt(len(new_list)))
    short_index = index - mult + 1
    
    for i in range(index * short_index):
        n = new_list[i] * new_list[i + index] * new_list[i + (2 * index)] * new_list[i + (3 * index)]
        if n > product:
            product = n
    
    for i in range(short_index):
        for j in range(index):
            n = new_list[i + (j * index)] * new_list[i + (j * index) + 1] * new_list[i + (j * index) + 2] * new_list[i + (j * index) + 3]
            if n > product:
                product = n
    
    for i in range(short_index):
        for j in range(short_index):
            n = new_list[i + (j * index)] * new_list[i + ((j + 1) * index) + 1] * new_list[i + ((j + 2) * index) + 2] * new_list[i + ((j + 3) * index) + 3]
            if n > product:
                product = n
    
    for i in range(short_index):
        for j in range(short_index):
            n = new_list[i + 3 + (j * index)] * new_list[i + 2 + ((j + 1) * index)] * new_list[i + 1 + ((j + 2) * index)] * new_list[i + ((j + 3) * index)]
            if n > product:
                product = n

    return product    

File: java/completed/euler0011.java
import lib.EulerTime;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

/**
 * Created by Holden on 11/20/2016. - COMPLETED
 *
 * Problem 11 - Largest Product in a Grid
 *
 */

public class Prob0011 {
    public static void main(String[] args) {
        long startTime = System.nanoTime();
        long answer = largestProduct("euler/src/x0011.txt", 4);
        long endTime = System.nanoTime();

        System.out.println(answer);

        long duration = (endTime - startTime);
        EulerTime.time(duration);
    }

    public static long largestProduct(String filename, int numbers) {
        int[][] grid = parseGrid(filename);
        int dimension = grid.length;
        int offset = numbers - 1;
        int largest = 0;

        // 4 numbers in a row
        for (int i = 0; i < dimension; i++) {
            for (int j = 0; j < (dimension - offset); j++) {
                int product = 1;
                for (int multiple = 0; multiple < numbers; multiple++) {
                    product *= grid[i][j+multiple];
                }

                if (product > largest) {
                    largest = product;
                }
            }
        }

        // 4 numbers in a column
        for (int i = 0; i < (dimension - offset); i++) {
            for (int j = 0; j < dimension; j++) {
                int product = 1;
                for (int multiple = 0; multiple < numbers; multiple++) {
                    product *= grid[i+multiple][j];
                }

                if (product > largest) {
                    largest = product;
                }
            }
        }

        // 4 numbers down diagonally
        for (int i = 0; i < (dimension - offset); i++) {
            for (int j = 0; j < (dimension - offset); j++) {
                int product = 1;
                for (int multiple = 0; multiple < numbers; multiple++) {
                    product *= grid[i+multiple][j+multiple];
                }

                if (product > largest) {
                    largest = product;
                }
            }
        }

        // 4 numbers up diagonally
        for (int i = offset; i < dimension; i++) {
            for (int j = 0; j < (dimension - offset); j++) {
                int product = 1;
                for (int multiple = 0; multiple < numbers; multiple++) {
                    product *= grid[i-multiple][j+multiple];
                }

                if (product > largest) {
                    largest = product;
                }
            }
        }

        return largest;
    }

    private static int[][] parseGrid(String filename) {
        int[][] grid;

        try {
            File file = new File(filename);
            FileReader fileReader = new FileReader(file);
            BufferedReader bufferedReader = new BufferedReader(fileReader);
            ArrayList<String> stringArray = new ArrayList<>();
            String line;
            while ((line = bufferedReader.readLine()) != null) {
                stringArray.add(line);
            }
            fileReader.close();

            int dimension = stringArray.size();
            grid = new int[dimension][dimension];

            for (int i = 0; i < dimension; i++) {
                String[] row = stringArray.get(i).split(" ");
                for (int j = 0; j < dimension; j++) {
                    grid[i][j] = Integer.parseInt(row[j]);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }

        return grid;
    }
}
